auth.js

```diff
@@ -0,0 +1,65 @@
+const { createSign } = require('crypto');
+
+const request = require('./request');
+
+const TOKEN_URL = 'https://oauth2.googleapis.com/token';
+const GRANT_TYPE = 'urn:ietf:params:oauth:grant-type:jwt-bearer';
+
+// When requesting a token, we specify 60 minutes. When setting the local
+// lifetime of a token, we specify 55 minutes. This simple mechanism ought to
+// work us around a race that could occur if we used the same number for both
+// values: our local lifetime check could pass, but still be rejected by
+// Google's API if we're operating right around the expiry epoch.
+//
+const EXP_SECS = 60 * 60;
+const LIFETIME_SECS = 60 * 55;
+
+const toB64 = (o) => Buffer.from(JSON.stringify(o)).toString('base64');
+const sign = (b, k) => createSign('RSA-SHA256').update(b).sign(k);
+
+const HEADER = toB64({ alg: 'RS256', typ: 'JWT' });
+
+const getJWT = ({ key, email, scopes }) => {
+  const now = Math.floor(Date.now() / 1000);
+  const payload = {
+    iss: email,
+    scope: scopes.join(' '),
+    aud: TOKEN_URL,
+    exp: now + EXP_SECS,
+    iat: now,
+  };
+
+  const unsigned = `${HEADER}.${toB64(payload)}`;
+  return `${unsigned}.${sign(unsigned, key).toString('base64')}`;
+};
+
+const getBearer = async ({ bearerCache, key, email, scopes }) => {
+  if (bearerCache.isWarm()) {
+    return bearerCache.value();
+  }
+
+  if (bearerCache.isFetching()) {
+    return bearerCache.wait();
+  }
+
+  bearerCache.setFetching();
+
+  try {
+    const res = await request({
+      url: TOKEN_URL,
+      method: 'POST',
+      form: {
+        grant_type: GRANT_TYPE,
+        assertion: getJWT({ key, email, scopes }),
+      },
+    });
+
+    bearerCache.setWarm(res.body.access_token, LIFETIME_SECS);
```

Comment:
Is there a chance that, despite requesting an `EXP_SECS` token, we might receive one with a shorter lifetime? Can the true expiration timestamp of the token be determined from the response? If both are true, it might avoid a class of bug to set the cache TTL here dynamically based on that.

Reply:
There is _definitely_ a chance that could happen, and in fact the response _does_ include the lifetime in seconds. Here's a sample:

```js
{ 
  access_token: <REDACTED>
  expires_in: 3599,
  token_type: 'Bearer' 
}
```

So: it's a great idea, and nice intuition. 

I'd be very surprised if the API changed suddenly to (A) actually respect the expiration argument and (B) returned a value _other_ than what we asked for. But, easy enough to use the value in the response... I'll update!

Reply:
Ah, great!